Shopify/cli
# Implement disabled prompt choices

### WHY are these changes introduced?

Fixes https://github.com/Shopify/internal-cli-foundations/issues/581

We want to communicate to the user better why certain extensions are not available.

### WHAT is this pull request doing?

Adding this functionality with the previous `SelectInput` implementation was very difficult so I've heavily refactored the state logic to:

- Use `useReducer` instead of `useState` to encapsulate it in an easier to understand reducer function.
- Not rely on `onChange` from `SelectPrompt` but instead pass an `onSubmit` function to `SelectInput`. This means that now `SelectInput` is in charge of handling the return key and the various shortcuts if `submitWithShortcuts` is true.
- `SelectInput` doesn't deal with indexes anymore but rather with a map of options that each have a `next` or `previous` properties. This way it's easy to check if the next or previous items are disabled or not.

Functionality-wise, I've removed the infinite scrolling that would reset the selected item to the top of the list when you reach the last item and press down arrow. Same for arrow up but in reverse.
I did this because if we want to properly paginate server side in the future we cannot support this behavior as pressin up arrow would try to fetch the last page, which is not something we can do with our GraphQL APIs AFAIK.

Example in `kitchen-sink prompts`:

![](https://screenshot.click/24-41-wcrdf-fc54k.gif)

### How to test your changes?

- Try creating an extension that has a limit of 1
- Run `extension generate` again and see that the option is greyed out and not selectable
